### PR TITLE
Enable acceleration setpoint based takeoff

### PR DESCRIPTION
**Pull request by Jaeyoung-Lim**
**Describe problem solved by this pull request**
When sending acceleration setpoints as part of offboard position setpoints as part of [SET_POSITION_TARGET_LOCAL_NED](https://mavlink.io/en/messages/common.html#SET_POSITION_TARGET_LOCAL_NED) mavlink message, the vehicle would not take off.

The reason seems to be that a downward acceleration was being copied while the vehicle was in `flying_but_ground_contact=true` state.  

**Describe your solution**
This PR adds the following.
- Enables a acceleration triggered takeoff based on the acceleration setpoint
- Fixes the takeoff logic so that the vehicle is able to take off while acceleration setpoints are being sent

**Test data / coverage**
Having acceleration setpoints do seem to improve the performance of trajectory tracking.
The below examples show a log of the vehicle tracking a circular trajectory with and without acceleration feed forward inputs.

- (BeforePR) Without Acceleration feedforward setpoints: https://review.px4.io/plot_app?log=3f8f3643-991b-435e-8286-289aa2920803
image

- (AfterPR) With Acceleration feedforward setpoints: https://review.px4.io/plot_app?log=7515c5b1-bfc1-4026-9a11-8eabda5134c0
image

**Additional context**
- Acceleration feedforward control was added in https://github.com/PX4/Firmware/pull/14212, but was not functional for offboard control.

### CODE REVIEW

**Review by mhkabir — APPROVED** [review]

**Review by MaEtUgR — CHANGES_REQUESTED** [review]
With this change no multicopter can land anymore!
It's a workaround and we need to find the root cause why ground contact stays with upwards acceleration.
Thanks @Jaeyoung-Lim for having a look at this!

**Comment by MaEtUgR** [comment]
After discussion we came to the conclusion that probably the vertical acceleration trips the takeoff ramp because it's designed based on velocity and I didn't have time to fully adapt it to the acceleration possibilities. We need to track that down and in the short term @Jaeyoung-Lim can try using only horizontal acceleration feed-forward. I'd like to have a very minimal example to run locall such that I can get in this state because I then expect it will be easy for me to find the problem.

File: src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp [hunk 1]
```diff
@@ -203,5 +203,12 @@ bool FlightTask::_checkTakeoff()
 		velocity_triggered_takeoff = _velocity_setpoint(2) < -0.3f;
 	}
 
-	return position_triggered_takeoff || velocity_triggered_takeoff;
+	// upwards acceleration setpoint
+	bool acceleration_triggered_takeoff = false;
+
+	if (PX4_ISFINITE(_acceleration_setpoint(2))) {
+		acceleration_triggered_takeoff = _acceleration_setpoint(2) < -0.3f;
```
Comment (dagar) [inline src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp:210]: Can you document this magic number?

**Comment by Jaeyoung-Lim** [comment]
@MaEtUgR I fixed the issue now, can you review again?

**Comment by MaEtUgR** [comment]
@Jaeyoung-Lim Good find! I did not find this when we were debugging before.
The main things to check are:
- Is the flag offboard only
- Is there a mode that erroneously sends acceleration that would take off assuming the base task only looks at velocity. Then we need to fix that task to not have unexpected takeoffs.
I'll quickly check.

**Comment by Jaeyoung-Lim** [comment]
> - Is the flag offboard only

Yes! Since the climbout mode was being set in offboard mode flags

> Is there a mode that erroneously sends acceleration that would take off assuming the base task only looks at velocity. Then we need to fix that task to not have unexpected takeoffs.

That is something I would like to check, but as of now I think the only way to use acceleration setpoints are through feedforward and not as a standalone.

It could be problematic if they don't set the acceleration ignore flag when sending `SET_TARGET_POSITION_LOCAL_NED` since if you set the `type_mask=0` it will enable all setpoints. However this is a problem in the mavlink spec so the user is responsible to use this safely.

**Comment by MaEtUgR** [comment]
> - Is the flag offboard only
> Yes! Since the climbout mode was being set in offboard mode flags

Got it, I double checked ✔️ 

> That is something I would like to check, but as of now I think the only way to use acceleration setpoints are through feedforward and not as a standalone.

Acceleration setpoints are supported internally as standalone, not just feed-forward. What's new with this pr is that you can actually trigger a takeoff with vertical acceleration setpoints only. I did not allow that before because I don't know what the velocity ramp ramps to. And I checked everything but the `SmoothVel` library to not publish any vertical acceleration setpoints that make it take off accidentally. After verifying also the `SmoothVel` library and the ramp we can merge.

**Comment by MaEtUgR** [comment]
Like I feared the acceleration-based takeoff without dedicated handling in the takeoff ramp breaks auto takeoff. It immediately seems to trigger a takeoff on arm, land again and then takes off again. Since I didn't understand why exactly this happens it could still be unrelated but I did a quick bisect test and it seems to be that. Let me find out why this happens.
image

**Comment by MaEtUgR** [comment]
I moved the important part into a separate pr such that we can move on with it:
https://github.com/PX4/PX4-Autopilot/pull/16498

**Comment by Jaeyoung-Lim** [comment]
@MaEtUgR Thanks!

**Comment by MaEtUgR** [comment]
I guess we do purely acceleration-based takeoffs in a different take then. We anyways discussed we need to move the takeoff logic into the tasks themselves such that they can decide when to do what and it should get easier to handle it then. In the meantime doing the takeoff by command or setting an upwards velocity setpoint to takeoff should do the trick.

**Pull request closed without merging**